apps/ewallet_db/lib/ewallet_db/transaction_request.ex

```diff
@@ -78,13 +106,148 @@ defmodule EWalletDB.TransactionRequest do
+      expiration_reason: "expired_request"
+    ])
+  end
+
+  @doc """
+  Gets a request with a "FOR UPDATE" lock on it. Should be called inside a transaction.
+  """
+  @spec get_with_lock(UUID.t) :: %TransactionRequest{} | nil
+  def get_with_lock(nil), do: nil
+  def get_with_lock(id) do
+    case Helpers.UUID.valid?(id) do
+      true ->
+        TransactionRequest
+        |> where([t], t.id == ^id)
+        |> lock("FOR UPDATE")
+        |> Repo.one()
+      false -> nil
+    end
+  end
+
+  @doc """
+  Touches a request by updating the `updated_at` field.
+  """
+  @spec get_with_lock(%TransactionRequest{}) :: {:ok, %TransactionRequest{}} |
+                                                {:error, Map.t}
+  def touch(request) do
+    request
+    |> touch_changeset(%{updated_at: NaiveDateTime.utc_now()})
+    |> Repo.update()
+  end
+
   @doc """
   Inserts a transaction request.
   """
   @spec insert(Map.t) :: {:ok, %TransactionRequest{}} | {:error, Map.t}
   def insert(attrs) do
     %TransactionRequest{}
-    |> create_changeset(attrs)
+    |> changeset(attrs)
     |> Repo.insert()
   end
+
+  @doc """
+  Updates a transaction request.
+  """
+  @spec insert(Map.t) :: {:ok, %TransactionRequest{}} | {:error, Map.t}
+  def update(%TransactionRequest{} = request, attrs) do
+    request
+    |> changeset(attrs)
+    |> Repo.update()
+  end
+
+  @spec valid?(%TransactionRequest{}) :: true | false
+  def valid?(request) do
+    request.status == @valid
+  end
+
+  @spec expired?(%TransactionRequest{}) :: true | false
+  def expired?(request) do
+    request.status == @expired
+  end
+
+  @spec expiration_from_lifetime(%TransactionRequest{}) :: NaiveDateTime.t | nil
+  def expiration_from_lifetime(request) do
+    lifetime? =
+      request.confirmable &&
+      request.consumption_lifetime &&
+      request.consumption_lifetime > 0
+
+    case lifetime? do
+      true  ->
+        NaiveDateTime.utc_now()
+        |> NaiveDateTime.add(request.consumption_lifetime, :millisecond)
+      _ -> nil
+    end
+  end
+
+  @doc """
+  Expires the given request with the specified reason.
+  """
+  @spec expire(%TransactionRequest{}) :: {:ok, %TransactionRequest{}} | {:error, Map.t}
+  def expire(request, reason \\ "expired_request") do
+    request
+    |> expire_changeset(%{
+      status: @expired,
+      expired_at: NaiveDateTime.utc_now(),
+      expiration_reason: reason
+    })
+    |> Repo.update()
+  end
+
+  @doc """
+  Expires the given request if the expiration date is past.
+  """
+  @spec expire_if_past_expiration_date(%TransactionRequest{}) ::
+                                       {:ok, %TransactionRequest{}} | {:error, Map.t}
+  def expire_if_past_expiration_date(request) do
+    expired? = request.expiration_date &&
+               NaiveDateTime.compare(request.expiration_date, NaiveDateTime.utc_now()) == :lt
+
+    case expired? do
+      true  -> expire(request)
+      _ -> touch(request)
```

Comment:
Answer: the touch is for the lock, if the record is locked in another process, it will have to wait to do the touch, therefore not proceeding with the consumptions count check